src/core/FileSearch.cpp

```diff
@@ -0,0 +1,102 @@
+/*
+ * FileSearch.cpp - File system search task
+ *
+ * Copyright (c) 2024 saker
+ *
+ * This file is part of LMMS - https://lmms.io
+ *
+ * This program is free software; you can redistribute it and/or
+ * modify it under the terms of the GNU General Public
+ * License as published by the Free Software Foundation; either
+ * version 2 of the License, or (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
+ * General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public
+ * License along with this program (see COPYING); if not, write to the
+ * Free Software Foundation, Inc., 51 Franklin Street, Fifth Floor,
+ * Boston, MA 02110-1301 USA.
+ *
+ */
+
+#include "FileSearch.h"
+
+#include <atomic>
+#include <chrono>
+#include <lmmsconfig.h>
+#include <ratio>
+
+#ifdef __MINGW32__
+#include <mingw.thread.h>
+#else
+#include <thread>
+#endif
+
+namespace lmms {
+FileSearch::FileSearch(const QString& filter, const QStringList& paths, const QStringList& extensions,
+	const QStringList& blacklist, QDir::Filters dirFilters, QDir::SortFlags sortFlags)
+	: m_filter(filter)
+	, m_paths(paths)
+	, m_extensions(extensions)
+	, m_blacklist(blacklist)
+	, m_dirFilters(dirFilters)
+	, m_sortFlags(sortFlags)
+{
+}
+
+void FileSearch::operator()()
+{
+	auto stack = QFileInfoList{};
+	for (const auto& path : m_paths)
+	{
+		if (m_blacklist.contains(path)) { continue; }
+
+		auto dir = QDir{path};
+		stack.append(dir.entryInfoList(m_dirFilters, m_sortFlags));
+
+		while (!stack.empty())
+		{
+			if (m_cancel.load(std::memory_order_relaxed)) { return; }
```

Comment:
Emitting `searchComplete`, even after cancelling, allows the `FileBrowser` to know that the search is done (regardless if it actually finished or not). This is necessary to reset the progress bar, and know when to destroy the search object.

Reply:
So we should emit searchComplete here instead of return immediately, right?

Reply:
As of right now, we are only cancelling the search when we are immediately ready to run a new one. Since we replace the current search task with the new one when enqueuing a new search, the old one gets destroyed (pretty much what `searchComplete` does but for when the search successfully completes, which we wouldn't know without that signal).

Since we only cancel when a new search is ready, `FileBrowser` effectively knows when the current search is "complete" (i.e., cancelled) since each new search that comes in should cancel the current one. When that happens, `FileBroswer` wouldn't depend on `searchComplete` to be emitted again to do any cleanup work.

```cpp
void FileBrowser::onSearch(const QString& filter)
{
	if (m_currentSearch)
	{
		m_currentSearch->cancel();
		m_currentSearch.reset();
	}
...
```

(I just noticed that we don't even have to be resetting the object here since the assignment that happens later using the new search will handle that for us).

Reply:
So it turns out we don't actually need to emit `searchComplete` here.

Reply:
Alright thanks for explaining! I didn't go that in depth with the logic there to see this. 👍